modules/dnn/src/layers/scatterND_layer.cpp

```diff
@@ -0,0 +1,263 @@
+        const size_t* ind_step = indices.step.p;
+
+        const int upd_ndims = updates.dims;
+        const int* upd_shape = updates.size.p;
+        const size_t* upd_step = updates.step.p;
+
+        typeDispatch(outputs[0].type(),
+                     ndims, shape, step, data.ptr<char>(),
+                     ind_ndims, ind_shape, ind_step, indices.ptr<char>(),
+                     upd_ndims, upd_shape, upd_step, updates.ptr<char>(),
+                     out.ptr<char>());
+    }
+
+    template<typename T, typename Functor>
+    void forward_impl(const Functor& rd,
+                      int ndims, const int* shape, const size_t* step, const char* input,
+                      int ind_ndims, const int* ind_shape, const size_t* ind_step, const char* indices,
+                      int upd_ndims, const int* upd_shape, const size_t* upd_step, const char* updates,
+                      char* out)
+    {
+        size_t input_total = 1;
+        for (int i = 0; i < ndims; i++)
+            input_total *= shape[i];
+        std::memcpy(out, input, input_total * sizeof(T));
+
+        int k = ind_shape[ind_ndims - 1]; // last dim of indices
+        size_t total_div_k = 1;
+        for (int i = 0; i < ind_ndims - 1; i++)
+            total_div_k *= ind_shape[i];
+
+        size_t updates_size = 1;
+        for (int i = ind_ndims - 1; i < upd_ndims; i++)
+            updates_size *= upd_shape[i];
+
+        size_t inp_start_offset = 0;
+        size_t ind_start_offset = 0;
+        size_t upd_start_offset = 0;
+        for (size_t i = 0; i < total_div_k; i++)
+        {
+            size_t t = i;
+            for (int j = ind_ndims - 2; j >= 0; j--)
+            {
+                size_t idx = t / ind_shape[j];
+                int offset_at_idx = (int)(t - idx * ind_shape[j]);
+                update_indices[j] = offset_at_idx;
+                t = idx;
+            }
+
+            ind_start_offset = 0;
+            for (int j = 0; j < ind_ndims - 1; j++)
+                ind_start_offset += update_indices[j] * ind_step[j];
+
+            const T* p_indices = (const T*)(indices + ind_start_offset);
+            std::fill(inp_current_indices.begin(), inp_current_indices.end(), 0);
+            for (int j = 0; j < k; j++)
+            {
+                inp_current_indices[j] = p_indices[j];
+            }
+
+            inp_start_offset = 0;
+            for (int j = 0; j < ndims; j++)
+            {
+                inp_start_offset += inp_current_indices[j] * step[j];
+            }
+
+            std::fill(upd_current_indices.begin(), upd_current_indices.end(), 0);
+            for (int j = 0; j < ind_ndims - 1; j++)
+            {
+                upd_current_indices[j] = update_indices[j];
+            }
+
+            upd_start_offset = 0;
+            for (int j = 0; j < upd_ndims; j++)
+            {
+                upd_start_offset += upd_current_indices[j] * upd_step[j];
+            }
```

Comment:
Do extra dimensions of indices matter?(not related to index tuple)
I'm not sure I got the docs correctly, but it seems that we can interpret like array of indices and then we have 1 loop for data index restoration and 1 loop for update?
I seem to be missing something, point me to the docs, please.

Reply:
The extra dimensions of indices are used to index indices (to index output) and updates. Do you mean flattening extra dimensions of indices. If yes, the extra dimensions of both indices and updates need to be flattened.

Reply:
Yes, do you think that's possible?

Reply:
I need to think of it carefully. Will reach back asap.

Reply:
I would further suggest to merge this as is, but open a ticket for a potential improvement, so that this PR doesn't block anyone.

Reply:
I prefer to keep current impl. I mean we can always index `indices` without `idx`, but I dont figure how to index `updates` without `idx`.

```python
# ScatterND impl in python
output = np.copy(data)
  update_indices = indices.shape[:-1]
  for idx in np.ndindex(update_indices):
      output[indices[idx]] = f(output[indices[idx]], updates[idx]) // how to index updates without idx?
```

Reply:
Thank you for your very insightful suggestion! Now the time of ScatterND has decreased half.

Reply:
Yay!